### PR TITLE
Change disabled text & icon color

### PR DESCRIPTION
**Pull request by FlyersPh9**
## Changes

Our disabled text color was not accessible ([not required by WCAG AA](https://www.w3.org/WAI/WCAG22/Understanding/non-text-contrast.html#inactive-controls)) but we wanted it to still be accessible.  But if we increased the contrast the color then became too similar to non-disabled items.  So we decreased the contrast of disabled text to give a bigger gap between disabled and enabled items & then introduced a disabled hover state to make it more accessible.

- Changed value of disabled text and disabled icon colors.
- Introduced new `--iui-color-text-disabled-hover` & `--iui-color-icon-disabled-hover` variables.
- Add hover state to disabled `.iui-field` and `.iui-list-item` to help with accessibility.

<details>
  <summary>Light themes</summary>

  *Scores against `iui-color-background-disabled` which should give the lowest contrast ratio score.*

| Theme | Disabled text before | Disabled text after | New disabled text hover |
| --- | --- | --- | --- |
| Light | `3.12` | `1.96` | `4.63` |
| Light high contrast | `3.68` | `3.27` | `7.5` |
</details>

<details>
  <summary>Dark themes</summary>

  *Scores against `iui-color-background` which should give the lowest contrast ratio score.*

| Theme | Disabled text before | Disabled text after | New disabled text hover |
| --- | --- | --- | --- |
| Dark | `2.93` | `2.93`  | `4.62` |
| Dark high contrast | `4.93` | `3.57` | `4.93` |
</details>

## Testing

- Visual tests.
- Checked [WCAG AA guidelines for inactive UI components](https://www.w3.org/WAI/WCAG22/Understanding/non-text-contrast.html#inactive-controls) to ensure compliance.

## Docs

Updated variables documentation page to list the new variables and explain the reasoning behind it.

### CODE REVIEW

File: packages/itwinui-variables/src/themes/light-hc.scss
Comment (FlyersPh9) [reply packages/itwinui-variables/src/themes/light-hc.scss]: That seems fair.  Added a new `$fg-750` to achieve this.

File: packages/itwinui-variables/src/themes/dark-hc.scss
Comment (FlyersPh9) [reply packages/itwinui-variables/src/themes/dark-hc.scss:145]: This is extremely opinion based, but I personally felt that the visual difference from default text to disabled text in dark theme was acceptable while the others could have the difference increased.

File: packages/itwinui-variables/src/themes/light.scss
Comment (mayank99) [reply packages/itwinui-variables/src/themes/light.scss:146]: Last week I had a thought: what if we decrease contrast for disabled items during resting state but we make it 4.5 when the disabled item is hovered or focused?

This would prevent an undesirable situation where the user is focusing an item but can't quite read the text inside it.

Comment (FlyersPh9) [reply packages/itwinui-variables/src/themes/light.scss:146]: We can try it.  I do wonder if that visual change will confuse a user to think the element is actionable / clickable.

Comment (FlyersPh9) [reply packages/itwinui-variables/src/themes/light.scss:146]: We'd need to introduce a new CSS variable, `--iui-color-text-disabled-hover`, right?

Comment (FlyersPh9) [reply packages/itwinui-variables/src/themes/light.scss:148]: Introduced `--iui-color-text-disabled-hover` & `--iui-color-icon-disabled-hover`, implemented only on `.iui-field` thus far.

Comment (mayank99) [reply packages/itwinui-variables/src/themes/light.scss:148]: We should do the same in other interactive components. MenuItem is a must, since that's where the original issue was reported.

Comment (FlyersPh9) [reply packages/itwinui-variables/src/themes/light.scss:148]: Had to fix #2192 first.  Since that merged, I've added `:hover` & `:focus-visible` to a list item that is disabled.  I did _not_ include `[data-iui-actionable='true']` because I am assuming if a list item is disabled it must also be actionable?  Not sure if that's a safe assumption or not.

Comment (mayank99) [reply packages/itwinui-variables/src/themes/light.scss:148]: It's not a safe assumption. `ListItem` is a low-level primitive so it can be used in any arbitrary ways. The `:hover` styling should only apply when it's actionable.

Comment (FlyersPh9) [reply packages/itwinui-variables/src/themes/light.scss:148]: Ok, more specifically targeting `[data-iui-actionable='true']` now.

Comment (FlyersPh9) [reply packages/itwinui-variables/src/themes/light.scss:148]: Should it also be targeting `:has(.iui-link-action)`?

Comment (FlyersPh9) [reply packages/itwinui-variables/src/themes/light.scss:148]: Try checking again, I increased the difference between disabled and disabled hover in dark theme.  The disabled hover state was not very noticeable only in dark mode.

Comment (mayank99) [reply packages/itwinui-variables/src/themes/light.scss:148]: Looks good now 👍

Could you also update the PR description and mention the new contrast ratios?

File: apps/css-workshop/backstop/results/bitmaps_reference/iTwinUI_button_Grouped_Borderless_hc_dark_0_demo-grouped-borderless_0_desktop.png
Comment (mayank99) [inline apps/css-workshop/backstop/results/bitmaps_reference/iTwinUI_button_Grouped_Borderless_hc_dark_0_demo-grouped-borderless_0_desktop.png:1]: Since there have been so many changes since the images were last updated, I think it'd be better to regenerate the images against `main`.

You can restore all images using these commands:
```
git checkout origin/main apps/css-workshop/backstop
```
```
git checkout origin/main apps/react-workshop/cypress-visual-screenshots
```

After that, just run the tests and approve images as normal, then repeat a couple times until tests pass.

File: packages/itwinui-variables/src/themes/light.scss
Comment (FlyersPh9) [reply packages/itwinui-variables/src/themes/light.scss:148]: Updated the changesets and PR description, will add %s in the coming minutes.

Comment (FlyersPh9) [reply packages/itwinui-variables/src/themes/light.scss:148]: While testing the contrast scores I noticed 2 things:
1. I was mainly testing against `iui-color-background-disabled` which gives the lowest score for light themes, but `iui-color-background` gives a lower score for dark themes.  Because of this our dark high contrast theme's disabled hover is not meeting 7.5.
2. Dark high contrast's muted text is scoring 7.48, 0.02 below the required minimum. 🫣

Comment (mayank99) [reply packages/itwinui-variables/src/themes/light.scss:148]: It's ok, because:
- [SC 1.4.6](https://www.w3.org/WAI/WCAG22/Understanding/contrast-enhanced.html) only requires 7:1, not 7.5:1. And disabled controls are specifically exempt.
- We are never going to pass all success criteria for AAA (it is near-impossible to do so). There is no such thing as "partial success"; it's either pass or fail. In other words, we can't make claims like "we are meeting AAA color contrast but not the other success criteria".

File: .changeset/cuddly-pumas-reply.md
Comment (mayank99) [inline .changeset/cuddly-pumas-reply.md:5]: 1. Changesets should be worded in past tense.
2. CSS/variables changesets (especially for visual changes) should have corresponding React changesets describing the high-level effects of the underlying changes.

Comment (mayank99) [reply .changeset/cuddly-pumas-reply.md:5]: Also can we enhance the documentation in some places to describe the outcome (and the rationale) of this PR? Developers should both understand the default behavior and be guided to do the right thing, to prevent end-user frustration ("why is this thing disabled?").

Comment (FlyersPh9) [reply .changeset/cuddly-pumas-reply.md]: 1. Updated to past tense.
2. Some day I'll remember this. 😅 Added React changeset.
3. Updated variables documentation to list the new variables with a very broad explanation of why. Also updated PR description.

**Review by mayank99 — APPROVED** [review]
LGTM!

I would also like to see more documentation at the component level, especially for disabled buttons and menu-items. But I won't block the PR on that.

File: .changeset/cuddly-pumas-reply.md
Comment (mayank99) [reply .changeset/cuddly-pumas-reply.md]: The React changesets should not mention implementation details (i.e. CSS classes). As mentioned above, they should describe "the high-level effects of the underlying changes" (i.e. what changed visually and in which components).

File: packages/itwinui-css/src/field.scss
Comment (mayank99) [reply packages/itwinui-css/src/field.scss:75]: This is because the react-workshop input doesn't have any text, it's only using placeholder.

File: .changeset/cuddly-pumas-reply.md
Comment (FlyersPh9) [reply .changeset/cuddly-pumas-reply.md]: Updated React changeset to:
> Added hover state to disabled button, select, input, textarea, and actionable list item to help with accessibility.

File: apps/css-workshop/backstop/results/bitmaps_reference/iTwinUI_input_Type_Default_light_0_demo-default_0_desktop.png
Comment (FlyersPh9) [reply apps/css-workshop/backstop/results/bitmaps_reference/iTwinUI_input_Type_Default_light_0_demo-default_0_desktop.png:1]: Even in production right now the disabled label & status message don't match, but they are similar enough to make that mistake.
Screenshot 2024-08-19 at 3 32 28 PM
Screenshot 2024-08-19 at 3 32 43 PM

**Pull request merged**
